### PR TITLE
Shader support multi shader pass

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Main modify:
- `Shader` add multi shader pass ability
- `BaseMaterial` add `setIsTransparent`、`setBlendMode`、`setRenderFace` method
- `PBRMaterial` can custom `shader` in constructor

Example:
 can see `PlanarShadowMaterial` in  https://github.com/oasis-engine/engine-toolkit/pull/71

### CODE REVIEW

File: packages/core/src/2d/text/TextRenderer.ts [hunk 1]
```diff
@@ -322,11 +322,21 @@ export class TextRenderer extends Renderer implements ICustomClone {
...
       const spriteElement = this._engine._spriteElementPool.getFromPool();
-      spriteElement.setValue(this, charRenderData.renderData, this.getMaterial(), charRenderData.texture);
+      spriteElement.setValue(
+        this,
+        charRenderData.renderData,
+        this.getMaterial(),
+        charRenderData.texture,
+        renderStates[0],
+        passes[0]
+      );
```
Comment (singlecoder) [inline packages/core/src/2d/text/TextRenderer.ts:339]: TextRenderer just run one pass if has multi-passes

File: packages/core/src/shader/Shader.ts [hunk 2]
```diff
@@ -123,17 +137,25 @@ export class Shader {
...
     this.name = name;
-    this._vertexSource = vertexSource;
-    this._fragmentSource = fragmentSource;
+
+    if (typeof vertexSourceOrShaderPasses === "string") {
+      this._passes.push(new ShaderPass(vertexSourceOrShaderPasses, fragmentSource));
+    } else {
+      for (let i = 0, n = vertexSourceOrShaderPasses.length; i < n; i++) {
+        this._passes.push(vertexSourceOrShaderPasses[i]);
+      }
```
Comment (singlecoder) [inline packages/core/src/shader/Shader.ts:157]: If vertexSourceOrShaderPasses.length is 0, will not has default pass

File: packages/core/src/material/Material.ts [hunk 3]
```diff
@@ -15,14 +15,47 @@ import { RenderQueueType } from "./enums/RenderQueueType";
 export class Material extends RefObject implements IClone {
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/material/Material.ts:15]: done

File: packages/core/src/shader/Shader.ts [hunk 4]
```diff
@@ -1,39 +1,50 @@
...
    * @param vertexSource - Vertex source code
    * @param fragmentSource - Fragment source code
    */
-  static create(name: string, vertexSource: string, fragmentSource: string): Shader {
+  static create(name: string, vertexSource: string, fragmentSource: string): Shader;
+
+  /**
+   * Create a shader.
+   * @param name - Name of the shader
+   * @param shaderPasses - Shader passe
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/shader/Shader.ts:38]: done

File: packages/core/src/Engine.ts [hunk 5]
```diff
@@ -344,8 +345,8 @@ export class Engine extends EventDispatcher {
   /**
    * @internal
    */
-  _getShaderProgramPool(shader: Shader): ShaderProgramPool {
-    const index = shader._shaderId;
+  _getShaderProgramPool(shader: ShaderPass): ShaderProgramPool {
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/Engine.ts:348]: done

File: packages/core/src/shader/Shader.ts (hunk 2 above)
Comment (GuoLei1990) [reply packages/core/src/shader/Shader.ts:163]: done

File: packages/core/src/material/BaseMaterial.ts [hunk 6]
```diff
@@ -38,19 +38,15 @@ export class BaseMaterial extends Material {
     if (value === this._isTransparent) return;
     this._isTransparent = value;
 
-    const {
-      depthState,
-      blendState: { targetBlendState }
-    } = this.renderState;
-
+    const renderState = this.renderState;
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/material/BaseMaterial.ts:41]: done

File: packages/core/src/shader/state/RenderState.ts [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import { Engine } from "../../Engine";
+import { RenderQueueType } from "../../material/enums/RenderQueueType";
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/shader/state/RenderState.ts]: done

**Review by singlecoder — APPROVED** [review]
+1

**Pull request merged**